core_lib/src/graphics/bitmap/bitmapbucket.cpp

```diff
@@ -0,0 +1,245 @@
+/*
+
+Pencil2D - Traditional Animation Software
+Copyright (C) 2005-2007 Patrick Corrieri & Pascal Naidon
+Copyright (C) 2012-2020 Matthew Chiawen Chang
+
+This program is free software; you can redistribute it and/or
+modify it under the terms of the GNU General Public License
+as published by the Free Software Foundation; version 2 of the License.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+*/
+#include "bitmapbucket.h"
+
+#include <QtMath>
+#include <QDebug>
+
+#include "editor.h"
+#include "layermanager.h"
+
+#include "layerbitmap.h"
+
+BitmapBucket::BitmapBucket()
+{
+}
+
+BitmapBucket::BitmapBucket(Editor* editor,
+                           QColor color,
+                           QRectF maxFillRegion,
+                           QPointF fillPoint,
+                           Properties properties):
+    mEditor(editor),
+    mBucketStartPoint(fillPoint),
+    mMaxFillRegion(maxFillRegion),
+    mProperties(properties)
+
+{
+    Layer* initialLayer = editor->layers()->currentLayer();
+    int initialLayerIndex = mEditor->currentLayerIndex();
+    int frameIndex = mEditor->currentFrame();
+
+    mBucketColor = qPremultiply(color.rgba());
+
+    mTargetFillToLayer = initialLayer;
+    mTargetFillToLayerIndex = initialLayerIndex;
+
+    if (properties.bucketFillToLayerMode == 1) {
+
+        auto result = findBitmapLayerBelow(initialLayer, initialLayerIndex);
+        mTargetFillToLayer = result.first;
+        mTargetFillToLayerIndex = result.second;
+    }
+
+    if (mTargetFillToLayer == nullptr) {
+        Q_ASSERT(true);
```

Comment:
Do you mean `Q_ASSERT(false)`? it's the case that should never happen, right?

Reply:
And looks like the `assert` shouldn't be put here as the `mTargetFillToLayer` could be null.

Reply:
I'm assuming based off of the changes you made that you realize that it can't be null. editor->layers()->currentLayer() asserts the return is not null, so initialLayer can't be null. findBitmapLayerBelow will return initialLayer or another layer which is allso asserted not to be null.

Reply:
Yes, I finally got it :p